core/src/main/java/org/apache/hadoop/fs/SparkGlobber.java

```diff
@@ -0,0 +1,293 @@
+          break;
+        }
+        if ((componentIdx < components.size() - 1) &&
+            (!globFilter.hasPattern())) {
+          // Optimization: if this is not the terminal path component, and we
+          // are not matching against a glob, assume that it exists.  If it
+          // doesn't exist, we'll find out later when resolving a later glob
+          // or the terminal path component.
+          for (FileStatus candidate : candidates) {
+            candidate.setPath(new Path(candidate.getPath(), component));
+          }
+          continue;
+        }
+        for (FileStatus candidate : candidates) {
+          if (globFilter.hasPattern()) {
+            FileStatus[] children = listStatus(candidate.getPath());
+            if (children.length == 1) {
+              // If we get back only one result, this could be either a listing
+              // of a directory with one entry, or it could reflect the fact
+              // that what we listed resolved to a file.
+              //
+              // Unfortunately, we can't just compare the returned paths to
+              // figure this out.  Consider the case where you have /a/b, where
+              // b is a symlink to "..".  In that case, listing /a/b will give
+              // back "/a/b" again.  If we just went by returned pathname, we'd
+              // incorrectly conclude that /a/b was a file and should not match
+              // /a/*/*.  So we use getFileStatus of the path we just listed to
+              // disambiguate.
+              if (!getFileStatus(candidate.getPath()).isDirectory()) {
+                continue;
+              }
+            }
+            for (FileStatus child : children) {
+              if (componentIdx < components.size() - 1) {
+                // Don't try to recurse into non-directories.  See HADOOP-10957.
+                if (!child.isDirectory()) continue;
+              }
+              // Set the child path based on the parent path.
+              child.setPath(new Path(candidate.getPath(),
+                  child.getPath().getName()));
+              if (globFilter.accept(child.getPath())) {
+                newCandidates.add(child);
+              }
+            }
+          } else {
+            // When dealing with non-glob components, use getFileStatus
+            // instead of listStatus.  This is an optimization, but it also
+            // is necessary for correctness in HDFS, since there are some
+            // special HDFS directories like .reserved and .snapshot that are
+            // not visible to listStatus, but which do exist.  (See HADOOP-9877)
+            FileStatus childStatus = getFileStatus(
+                new Path(candidate.getPath(), component));
+            if (childStatus != null) {
+              newCandidates.add(childStatus);
+            }
+          }
+        }
+        candidates = newCandidates;
+        // When size of candidates has reached the threshold, stop expanding other
```

Comment:
This file folk org.apache.hadoop.fs.Globber from hadoop-common-2.7.2, I only add the logic from line 257 to 266 which control the expanding logic for candidates dirs.

Reply:
IMHO I just a bit worry about maintainability about this class...  btw, there is no test for this class in hadoop-common?

Reply:
There is test for `hadoop.fs.Globber` in [TestGlobExpander.java](https://github.com/apache/hadoop/blob/trunk/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/fs/TestGlobExpander.java) and [TestGlobPattern.java](https://github.com/apache/hadoop/blob/trunk/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/fs/TestGlobPattern.java).

Reply:
If we modify this class in the spark-side in future, how we should check if these tests passed?

Reply:
Is it also necessary port the UT? I think maybe enough comment and limit change is enough cause UT still has maintenance cost. What's your opinion?

Reply:
cc: @cloud-fan